Add Tutorials `6. Store Encrypted Data` and `7. Fetch Encrypted Data` (#294)
iotaledger/iota.py: docs/tutorials.rst

should be decremented by 1i, and you should see a new address, which was
actually the ``change_address``.

6. Store Encrypted Data
-----------------------

In this example, you will learn how to:

- **Convert Python data structures to JSON format.**
- **Encrypt data and include it in a zero-value transaction.**
- **Generate an address with a PyOTA specific utility tool.**
- **Store the zero-value transaction with encrypted data on the Tangle.**

.. warning::

    We will use the ``simple-crypt`` external library for encryption/decryption.
    Before proceeding to the tutorial, make sure you install it by running::

        pip install simple-crypt

Code
~~~~
.. literalinclude:: ../examples/tutorials/06_store_encrypted.py
   :linenos:

Discussion
~~~~~~~~~~
.. literalinclude:: ../examples/tutorials/06_store_encrypted.py
   :lines: 1-15
   :lineno-start: 1

The first odd thing we notice among the imports is
:py:class:`~iota.crypto.addresses.AddressGenerator`. It is a PyOTA specific
utility that generates addresses given your seed.

> We will use the ``encrypt`` method to enchiper the data, and ``b64encode`` for

Reviewer: ```suggestion
We will use the ``encrypt`` method to encipher the data, and ``b64encode`` for
```

   :lines: 25-26
   :lineno-start: 25

Notice, that ``data`` is a Python ``dict`` object. As a common way of exchanging
data on the web, we would like to convert it to JSON format. The ``json.dumps()``
method does exactly that, and the result is a JSON formatted plaintext.

.. literalinclude:: ../examples/tutorials/06_store_encrypted.py
   :lines: 28-37
   :lineno-start: 28

Next, we will encrypt this data with a secret password we obtain from the user.

.. note::

    When you run this example, please remember the password at least until the
    next tutorial!

The output of the ``encrypt`` method is a ``bytes`` object in Python3 and
contains many special characters. This is a problem, since we can only convert
ASCII characters from ``bytes`` directly into :py:class:`TryteString`.

Therefore, we first encode our binary data into ASCII characters with `Base64`_
encoding.

.. literalinclude:: ../examples/tutorials/06_store_encrypted.py
   :lines: 39-57
   :lineno-start: 39

Now, we are ready to construct the transfer. We convert the encrypted `Base64`_
encoded data to trytes and assign it to the :py:class:`ProposedTransaction`

> object's message argument.

Reviewer: ```suggestion
object's ``message`` argument.
```

An address is also needed, so we generate one with the help of
:py:class:`~iota.crypto.addresses.AddressGenerator` and its
:py:meth:`~iota.crypto.addresses.AddressGenerator.get_addresses` method. Feel

> free to chose the index of the generated address, and don't forget, that the

Reviewer: ```suggestion
free to choose the index of the generated address, and don't forget, that the
```

    Before proceeding to the tutorial, make sure you install it by running::

        pip install simple-crypt

Code
~~~~
.. literalinclude:: ../examples/tutorials/07_fetch_encrypted.py
   :linenos:

Discussion
~~~~~~~~~~
.. literalinclude:: ../examples/tutorials/07_fetch_encrypted.py
   :lines: 1-14
   :lineno-start: 1

In contrast to `6. Store Encrypted Data`_ where we intended to encrypt data, in
this tutorial we will do the reverse, and decrypt data from the Tangle.
Therefore, we need the ``decrypt`` method from ``simplecrypt`` library and the
``b64decode`` method from ``base64`` library.

Furthermore, ``getpass`` is needed to prompt the user for a decryption
password, and ``json`` for deserializing JSON formatted string into Python
object.

.. literalinclude:: ../examples/tutorials/07_fetch_encrypted.py
   :lines: 16-17
   :lineno-start: 16

To fetch transactions or bundles from the Tangle, a reference is required to
retreive them from the network. Transactions are identified by their
transaction hash, while a group of transaction (a bundle) by bundle hash.
Hashes ensure the integrity of the Tangle, since they contain verifiable

> information on the content of the transfer objects.

Reviewer: ```suggestion
information about the content of the transfer objects.
```

``input()`` asks the user to give the tail transaction hash of the bundle
that holds the encrypted messages. The tail transaction is the first in the
bundle with index 0. Copy and paste the tail transaction hash from the console
output of `6. Store Encrypted Data`_ when prompted.

.. literalinclude:: ../examples/tutorials/07_fetch_encrypted.py
   :lines: 19-21
   :lineno-start: 19

Next, we fetch the bundle from the Tangle with the help of the
:py:meth:`~Iota.get_bundles` extended API command. It takes a list of tail
transaction hashes and returns the bundles for each of them. The response
``dict`` contains a ``bundles`` key with the value being a list of bundles
in the same order as the input argument hashes. Also note, that the bundles
in the response are actual PyOTA :py:class:`Bundle` objects.

To simplify the code, several operations are happening on line 21:

- Calling :py:meth:`~Iota.get_bundles` that returns a ``dict``,
- accessing the ``'bundles'`` key in the ``dict``,

> - and taking the first element of the the list if bundles in the value

Reviewer: ```suggestion
- and taking the first element of the the list of bundles in the value
```

  associated with the key.

.. literalinclude:: ../examples/tutorials/07_fetch_encrypted.py
   :lines: 23-39
   :lineno-start: 23

The next step is to extract the content of the message fields of the
transactions in the bundle. We call :py:meth:`Bundle.get_messages` to carry
out this operation. The method returns a list of unicode strings, essentially
the ``signature_message_fragment`` fields of the transactions, decoded from
trytes into unicode characters.

We then combine these message chunks into one stream of characters by using

> ``sting.join()``.

Reviewer: ```suggestion
``string.join()``.
```

We know that at this stage that we can't make sense of our message, because it
is encrypted and encoded into `Base64`_. Let's peel that onion layer by layer:

- On line 28, we decode the message into bytes with ``b64decode``.

> - On line 31, we ask the user for a decryption password.

Reviewer: ```suggestion
- On line 31, we ask the user for the decryption password (from the previous tutorial).
```
